src/frontend/Ast_to_Mir.ml

```diff
@@ -770,11 +763,15 @@ let%expect_test "read data" =
      (For (loopvar sym1__) (lower (Lit Int 1)) (upper (Lit Int 5))
       (body
        (Block
-        ((Assignment (mat ((Single (Var sym1__))))
-          (Indexed
-           (FunApp CompilerInternal FnReadData__
-            ((Lit Str mat) (Lit Str matrix) (Lit Int 10) (Lit Int 20)))
-           ((Single (Var sym1__)))))))))) |}]
+        ((For (loopvar sym2__) (lower (Lit Int 1))
+          (upper (FunApp StanLib Times__ ((Lit Int 10) (Lit Int 20))))
+          (body
+           (Block
+            ((Assignment (mat ((Single (Var sym1__)) (Single (Var sym2__))))
+              (Indexed
+               (FunApp CompilerInternal FnReadData__
+                ((Lit Str mat) (Lit Str matrix) (Lit Int 10) (Lit Int 20)))
+               ((Single (Var sym1__)) (Single (Var sym2__)))))))))))))) |}]
 
 let%expect_test "read param" =
   let m = mir_from_string "parameters { matrix<lower=0>[10, 20] mat[5]; }" in
```

Comment:
In this test, the way you are assigning to the array `mat` seems very inefficient, or am I wrong? It seems to me like you are doing something like
```
for (i in 1:n)
  x[i] = do_possibly_expensive_computation()[i]
```
Is that how you always assign to arrays? Or is this specific to the read_param situation.

Reply:
General purpose array assignment isn't simplified at all in the MIR yet, so it does the same thing stan2 does (using the `stan::model::assign` function). And here we know that reading parameters isn't expensive, and I am tentatively okay with punting that to code-gen for the backends where there is a more efficient way to do this (but open to other thoughts, just seems secondary at the moment).